### ISSUE TITLE
Incorrect handling of DST changeover for `PrimitiveDateTime`

### ISSUE THREAD

**Issue by korrat** [Yuri6037/time-tz#5]
When using `assume_timezone` on a `PrimitiveDateTime` close to the forward DST changeover, I get an unexpected result for the offset in the end. In my opinion, `time-tz` does not yet handle the timezone correctly in this situation.

The following test demonstrates the issue. I've also included a test for the backward changeover, but I think the behaviour there is acceptable.

```rust
mod time_tz {
    use time::macros::datetime;
    use time_tz::timezones::db::CET;
    use time_tz::PrimitiveDateTimeExt as _;

    #[test]
    fn handles_forward_changeover() {
        assert_eq!(
            datetime!(2022-03-27 01:30).assume_timezone(CET),
            datetime!(2022-03-27 01:30 +01:00)
        );
    }

    #[test]
    fn handles_backward_changeover() {
        // During backward changeover, the hour between 02:00 and 03:00 occurs twice, so either answer is correct
        /* assert_eq!(
            datetime!(2022-10-30 02:30).assume_timezone(CET),
            datetime!(2022-10-30 02:30 +02:00)
        ); */
        assert_eq!(
            datetime!(2022-10-30 02:30).assume_timezone(CET),
            datetime!(2022-10-30 02:30 +01:00)
        );
    }
}
```

On 2022-03-27, the date of DST changeover in CET, 01:30 in CET is 00:30 in UTC, but 01:30 in UTC is 03:30 in CET due to DST changeover.

My expectation of `assume_timezone` was that it behaves similar to [`assume_offset`](https://docs.rs/time/latest/time/struct.PrimitiveDateTime.html#method.assume_offset) on `PrimitiveDateTime`, which considers the `PrimitiveDateTime` to be in the provided offset.

The exhibited behaviour is correct when treating the `PrimitiveDateTime` as UTC, but I'd argue that this would not be very useful. If this is indeed the expected behaviour, the documentation should highlight this fact.

I think the issue comes down to [this line in `assume_timezone`](https://github.com/Yuri6037/time-tz/blob/59412aced942074f758aa5538fa0b9af8c386840/src/lib.rs#L53):

```rust
let offset = tz.get_offset_utc(&self.assume_utc());
```

---

For completeness, I've included the behaviour of `chrono-tz`. Note that the second test panics, since the local time 02:30 is ambiguous during backward DST changeover in CET.

```rust
mod chrono {
    use chrono::NaiveDate;
    use chrono::NaiveDateTime;
    use chrono::NaiveTime;
    use chrono::TimeZone;
    use chrono_tz::CET;

    #[test]
    fn handles_forward_changeover() {
        let local = CET
            .from_local_datetime(&NaiveDateTime::new(
                NaiveDate::from_ymd(2022, 3, 27),
                NaiveTime::from_hms(01, 30, 00),
            ))
            .unwrap();

        assert_eq!(local, CET.ymd(2022, 3, 27).and_hms(01, 30, 00));
    }

    #[test]
    #[should_panic]
    fn handles_backward_changeover() {
        let local = CET
            .from_local_datetime(&NaiveDateTime::new(
                NaiveDate::from_ymd(2022, 10, 30),
                NaiveTime::from_hms(02, 30, 00),
            ))
            .unwrap();

        assert_eq!(local, CET.ymd(2022, 10, 30).and_hms(02, 30, 00));
        assert_eq!(local, CET.ymd(2022, 10, 30).and_hms(02, 30, 00));
    }
}
```

**Comment by Yuri6037** [comment]
Well it does not behave like assume_offset. It assumes the PrimitiveDateTime is in UTC. That system has not been designed to work on non-UTC date times. You should first convert all date times to UTC.

**Comment by Yuri6037** [comment]
@korrat I've just pushed a new version 0.7.0 in master (not yet in crates.io). That version includes a breaking change from 0.6 which uses the default offset for the given timezone in assume_timezone. This works with your test case and should work on any PrimitiveDateTime which is proper for the current timezone.

Does this new version 0.7 in master behave like you'd expect?

Regarding chrono-tz, unfortunately I can't achieve that much details with a `PrimitiveDateTime` because a `PrimitiveDateTime` is not convertible to a unix timestamp, so it's not possible to perform binary search directly on the `PrimitiveDateTime`. To provide ambiguity checking a new API function would have to be provided but on `OffsetDateTime` which can be converted to a unix timestamp.

**Comment by korrat** [comment]
Thanks, @Yuri6037. While version 0.7 indeed behaves as I'd expect for that one test case, it fails two other test cases now: 

* for a local date time of 02:30, which is invalid in CET, the result is `2022-03-27T02:30+01:00`.
* for a local date time of 03:30, which is after DST changeover, the result is `2022-03-27T03:30+01:00`, but should be `2022-03-27T03:30+02:00`

For naming, you might want to change the name of the method to something like `assume_timezone_local`, to prevent silent changes for people relying on the UTC offset behaviour now.

---
As for chrono-tz, it seems that `NaiveDateTime::timestamp` simply assumes UTC offset (https://docs.rs/chrono/latest/chrono/naive/struct.NaiveDateTime.html#method.timestamp) and chrono-tz adjusts the effective timestamps through some adjustment in [`FixedTimespanSet::local_span`](https://github.com/chronotope/chrono-tz/blob/46ac4118358c0adc6e2601c6964e88188d015522/src/timezone_impl.rs#L228). I don't understand why that works yet.

**Comment by Yuri6037** [comment]
> Thanks, @Yuri6037. While version 0.7 indeed behaves as I'd expect for that one test case, it fails two other test cases now:
> 
> * for a local date time of 02:30, which is invalid in CET, the result is `2022-03-27T02:30+01:00`.
> * for a local date time of 03:30, which is after DST changeover, the result is `2022-03-27T03:30+01:00`, but should be `2022-03-27T03:30+02:00`
> 
> For naming, you might want to change the name of the method to something like `assume_timezone_local`, to prevent silent changes for people relying on the UTC offset behaviour now.
> 
> As for chrono-tz, it seems that `NaiveDateTime::timestamp` simply assumes UTC offset (https://docs.rs/chrono/latest/chrono/naive/struct.NaiveDateTime.html#method.timestamp) and chrono-tz adjusts the effective timestamps through some adjustment in [`FixedTimespanSet::local_span`](https://github.com/chronotope/chrono-tz/blob/46ac4118358c0adc6e2601c6964e88188d015522/src/timezone_impl.rs#L228). I don't understand why that works yet.

Thank you, I'll see what I can do. About the naming, it won't be a problem since this will be published under a new 1.0 release which should be completely independent from 0.6 (which is already issuing deprecation warning).

**Comment by Yuri6037** [comment]
@korrat I've now included all necessary changes and also, for the sake of completeness, I've added ambiguity checking. 

All test cases you gave me are now passing.

**Comment by korrat** [comment]
Can confirm, all cases I can think of are handled nicely. Thanks a lot for being so quick to fix this!

**Comment by Yuri6037** [comment]
The new version with all these fixes is now released under 1.0.0 on [crates.io](https://crates.io/crates/time-tz).
